quinn-proto/src/cid_generator.rs

```diff
@@ -0,0 +1,54 @@
+use rand::RngCore;
+
+use crate::shared::ConnectionId;
+use crate::MAX_CID_SIZE;
+
+/// Generates connection IDs for incoming connections
+pub trait ConnectionIdGenerator: Send {
+    /// Generates a new CID
+    fn generate_cid(&mut self) -> ConnectionId;
+    /// Performs any validation if it is needed (e.g. HMAC, etc)
+    fn validate_cid(&mut self, cid: &ConnectionId) -> bool;
```

Comment:
It looks like this is only used to skip checking an `Endpoint`'s hash table for a CID, but that hash table should return `None` for invalid CIDs anyway, because we'd never have inserted them, so naively we could do without it. Is the intention to guard against timing attacks when a cryptographic mechanism is in use? If so, that should be noted in the docs so future readers aren't confused about the benefit.

Reply:
> It looks like this is only used to skip checking an Endpoint's hash table for a CID, but that hash table should return None for invalid CIDs anyway, because we'd never have inserted them, so naively we could do without it.
Yes I have noticed that hash table will guarantee that non-existing CID will remain invalid. The purpose of validating CID is indeed based on some security concern. 

The major purpose of this API is to limit the amount of time for which a CID is valid in order to reduce the number of valid IDs that could be accumulated by an attacker. I am pushing another PR later (where peer actively sends NEW_CONNECTION_ID frame to rotate the CID). As a result, `fn validate_cid` is used to mark any CID that may still appear in hash table but considered invalid by application layer logic. 

This seems like an optional API. Thus, I am providing a default implementation and more docs.

Reply:
And I think timing attack is a good example on what may happen if an attacker holds a number of valid CIDs for a long time.